Sources/PointFree/Routes.swift

```diff
@@ -86,11 +90,47 @@ private func redirectUnrelatedDomains<A>(
         } else {
           return nil
         }
-      }
-      ?? middleware(conn)
+        }
+        ?? middleware(conn)
+    }
+}
+
+// TODO: move to HttpPipeline
+private func requireHttps<A>(
+  _ middleware: @escaping Middleware<StatusLineOpen, ResponseEnded, A, Data?>
+  )
+  -> Middleware<StatusLineOpen, ResponseEnded, A, Data?> {
+
+    return { conn in
+      conn.request.url
+        .filterOptional { $0.scheme == .some("http") }
+        .flatMap(makeHttps)
+        .map {
+          conn
+            |> writeStatus(.movedPermanently)
+            |> writeHeader(.location($0.absoluteString))
+            |> map(const(nil))
+            |> closeHeaders
+            |> end
+        }
+        ?? middleware(conn)
     }
 }
 
+// TODO: move to httppipeline?
+private func makeHttps(url: URL) -> URL? {
+  var components = URLComponents(url: url, resolvingAgainstBaseURL: false)
+  components?.scheme = "https"
+  return components?.url
+}
+
+// TODO: move to prelude
+extension Optional {
+  fileprivate func filterOptional(_ p: (Wrapped) -> Bool) -> Optional {
```

Comment:
we should prob move this to prelude at some point. it's an optional specific version of `mfilter: (a -> Bool) -> m a -> m a`

Reply:
Definitely! I was tempted to just call it `filter` since we have it on `Event` and `Array` that way, but `filterOptional` was a lot easier to read/understand here on GitHub.

Reply:
yeah def, I'm down to do just `filter` too.

or to align with `mfilter` maybe it should be `optionalFilter`.